prov/sockets/src/sock_progress.c

```diff
@@ -2150,15 +2152,30 @@ static int sock_pe_progress_rx_pe_entry(struct sock_pe *pe,
 					struct sock_pe_entry *pe_entry,
 					struct sock_rx_ctx *rx_ctx)
 {
-	int ret;
+	int ret, i;
 
 	if (sock_comm_is_disconnected(pe_entry)) {
 		SOCK_LOG_DBG("conn disconnected: removing fd from pollset\n");
-		sock_epoll_del(&pe_entry->ep_attr->cmap.epoll_set,
-			       pe_entry->conn->sock_fd);
 		sock_pe_poll_del(pe, pe_entry->conn->sock_fd);
 		ofi_close_socket(pe_entry->conn->sock_fd);
 
+		fastlock_acquire(&pe_entry->ep_attr->lock);
+		sock_epoll_del(&pe_entry->ep_attr->cmap.epoll_set,
+			       pe_entry->conn->sock_fd);
+		if (idm_lookup(&pe_entry->ep_attr->conn_idm, pe_entry->conn->sock_fd))
+			idm_clear(&pe_entry->ep_attr->conn_idm, pe_entry->conn->sock_fd);
+		
+		for (i = 0 ; i < pe_entry->ep_attr->cmap.used; i++) {
+			if (pe_entry->ep_attr->cmap.table[i].sock_fd == pe_entry->conn->sock_fd) {
+				pe_entry->ep_attr->cmap.table[i].ep_attr = NULL;
+				pe_entry->ep_attr->cmap.table[i].sock_fd = 0;
+				pe_entry->ep_attr->cmap.table[i].address_published = 0;
+				pe_entry->ep_attr->cmap.table[i].connected = 0;
+			}
+		}
+		pe_entry->ep_attr->cmap.used--;
+		fastlock_release(&pe_entry->ep_attr->lock);
```

Comment:
We can move this piece of code into a separate function, say sock_conn_release() or something.